internal/controller/pkg/runtime/reconciler.go

```diff
@@ -0,0 +1,468 @@
+		// Wait for signature verification to complete before proceeding.
+		if cond := pr.GetCondition(v1.TypeVerified); cond.Status != corev1.ConditionTrue {
+			log.Debug("Waiting for signature verification controller to complete verification.", "condition", cond)
+			// Initialize the healthy condition if they are not already set to
+			// communicate the status of the package.
+			if pr.GetCondition(v1.TypeHealthy).Status == corev1.ConditionUnknown {
+				status.MarkConditions(v1.Unhealthy().WithMessage("Waiting for signature verification to complete"))
+				return reconcile.Result{}, errors.Wrap(r.client.Status().Update(ctx, pr), "cannot update status with awaiting verification")
+			}
+			return reconcile.Result{}, nil
+		}
+	}
+
+	var pullSecretFromConfig string
+	// Read applied image config for SetImagePullSecret from the package
+	// revision status, so that we can use the same pull secret without having
+	// to resolve it again.
+	for _, icr := range pr.GetAppliedImageConfigRefs() {
+		if icr.Reason == v1.ImageConfigReasonSetPullSecret {
+			// Get applied image config to find the pull secret.
+			ic := &v1beta1.ImageConfig{}
+			if err := r.client.Get(ctx, types.NamespacedName{Name: icr.Name}, ic); err != nil {
+				err = errors.Wrap(err, errGetPullConfig)
+				status.MarkConditions(v1.Unhealthy().WithMessage(err.Error()))
+				_ = r.client.Status().Update(ctx, pr)
+				r.record.Event(pr, event.Warning(reasonImageConfig, err))
+				return reconcile.Result{}, err
+			}
+			pullSecretFromConfig = ic.Spec.Registry.Authentication.PullSecretRef.Name
+			break
+		}
+	}
+
+	// Initialize the runtime manifest builder with the package revision
+	opts, err := r.builderOptions(ctx, pr)
+	if err != nil {
+		log.Debug(errManifestBuilderOptions, "error", err)
+		err = errors.Wrap(err, errManifestBuilderOptions)
+		status.MarkConditions(v1.Unhealthy().WithMessage(err.Error()))
+		_ = r.client.Status().Update(ctx, pr)
+		r.record.Event(pr, event.Warning(reasonSync, err))
+		return reconcile.Result{}, err
+	}
+	if pullSecretFromConfig != "" {
+		opts = append(opts, BuilderWithPullSecrets(pullSecretFromConfig))
+	}
+	builder := NewDeploymentRuntimeBuilder(pr, r.namespace, opts...)
+
+	// Deactivate revision if it is inactive.
+	if pr.GetDesiredState() == v1.PackageRevisionInactive {
+		if err := r.runtimeHook.Deactivate(ctx, pr, builder); err != nil {
+			err := errors.Wrap(err, "failed to run deactivation hook")
+			r.log.Info("Error", "error", err)
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{Requeue: false}, nil
+	}
+
+	// Migrate provider deployment selector, if needed.
+	// Note(turkenh): This migration could be removed in the future when
+	// we no longer support older Crossplane versions that use the old
+	// provider deployment selector labels. The latest Crossplane version
+	// using the old selector labels is v1.20.0.
+	if err := r.migrateProviderDeploymentSelectors(ctx, pr, builder); err != nil {
```

Comment:
I found this method name confusing. It doesn't really migrate the deployment, right? It looks like it just deleted old-style deployments so that the hooks can create a new-style one.

That said, not sure what a better name would be. `deleteLegacyDeployment` maybe?